### ISSUE TITLE
Secret service integration in 2.7

### ISSUE THREAD

**Issue by nod0n** [keepassxreboot/keepassxc#7464]
## Overview
I'm testing 2.7 as appimage. I use the secrect service integration feature. When I open Remmina (a GTK RDP software), I have to every time allow Remmina to access it's entries. This happens, even though the "remember" checkbox is enabled. I don't know if this happens for all Remmina entries or if it has to do something with my database, but 

### setup secret service
1. Create an empty test database
2. enable secret service in the settings
3. expose a newly created folder, in the new test database, to the secret service
### Remmina
1. install remmina (if you don't have it already)
2. create a new RDP entry with a user and password.
3. quit remmina
4. open remmina
5. allow access to the entry in the pop up window with the "remember" button enabled
6. quit remmina
7. open remmina

## Expected Behavior
As I allowed access to the entry with the remember button enabled, I would expect that I don't get asked again, if remmina is allowed to access the entry.

## Actual Behavior
I have to allow the access to each remmina entry with a password, every time I open remmina.

## Context
KeePassXC - Version 2.7.0-snapshot
Build Type: Snapshot
Revision: d10c5a4
Distribution: AppImage

Qt 5.15.2
Debugging mode is disabled.

Operating system: KDE neon User - 5.24
CPU architecture: x86_64
Kernel: linux 5.13.0-30-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare
- YubiKey
- Secret Service Integration

Cryptographic libraries:
- Botan 2.12.1

Operating System: Linux (KDE Neon which is Ubuntu 20.04 with KDE apt repos)
Desktop Env: Plasma (KDE)
Windowing System: X11

**Comment by nod0n** [comment]
Starting Element, has the same behavior, but it only asks for one single entry, not for 10-20 like Remmina. This problem is not specific to Remmina, just with Remmina it is more obvious, because it reads multiple entries in KeePassXC.

**Comment by droidmonkey** [comment]
@Aetf it appears that DBusClient::setAllAuthorized is never called. Can you take a look into this? https://github.com/keepassxreboot/keepassxc/blob/4f0710350fb53c25db6824f96961d8d17a29347c/src/fdosecrets/dbus/DBusClient.cpp#L135-L138

**Comment by Aetf** [comment]
@nod0n, I suppose those are different entries asked by Remmnia each time?

The remember check means there won't be another prompt when the *same* entry is requested by the client.

@droidmonkey, you are right that `DBusClient:setAllAuthorized` is never called.
I didn't wire up that to the prompt dialog because I couldn't find a way to put that in the dialog in a non-confusing way, given that we have to ask again when the client restarts anyway.

Here are all possible combinations of actions. If we can figure out what should be included in the dialog I can prepare
a PR.

| Action | Selected entries | All listed entries | All entries including future ones until the client stops |
|:-:|:-:|:-:|:-:|
|Deny| - | Current "Deny All" button + checked remember | - |
|Allow| Current "Allow Selected" button + checked remember | (1) | (2) |
|Deny once| - | "Deny All" button | Doesn't make sense |
|Allow once| "Allow Selected" button | (1) | Doesn't make sense |

(1) can be achieved by selecting all listed entries, and I think @nod0n wants (2).

Note that it is not possible in KeePassXC only to identify clients once their process is gone, so it's not possible to implement the remember concept for all and future accesses which we are familiar with in other applications.

**Comment by droidmonkey** [comment]
Perhaps a simple "Allow all for client"

**Comment by michaelk83** [comment]
> Note that it is not possible in KeePassXC only to identify clients once their process is gone, so it's not possible to implement the remember concept for all and future accesses

You'd have to rely on the detected binary path. You may add fingerprinting for extra confidence that it's really the same binary that was previously authorized (#6458).

**Comment by droidmonkey** [comment]
I'm still more in favor of the "Remember only while keepassxc is running" approach rather then baking that memory into the database. Upon system restart or restart of keepassxc you should be reestablishing the trust relationships with the dbus clients.

**Comment by Aetf** [comment]
> You'd have to rely on the detected binary path. You may add fingerprinting for extra confidence that it's really the same binary that was previously authorized (#6458).

As mentioned in #6458, there are still open questions regarding client fingerprinting. So before we can figure
something out, I'm assuming it's not possible yet.

> I'm still more in favor of the "Remember only while keepassxc is running" approach rather than baking that memory into the database. Upon system restart or restart of keepassxc you should be reestablishing the trust relationships with the dbus clients.

This is orthogonal to whether to "remember only while client is running". I don't have a strong opinion on this one.

> Perhaps a simple "Allow all for client"

Something like the following?
What is the relationship between "remember" and "Allow All"? I suppose "Allow All" sort of implies  "remember"?

```
Client xxx is requesting access to the following entries:
------------------------------------------------------------
- process tree
  + parent process
    + child process
------------------------------------------------------------
* aaa
* bbb
------------------------------------------------------------
[ ] remember         [Deny All] [Allow Selected] [Allow All]
```

**Comment by michaelk83** [comment]
> before we can figure something out, I'm assuming it's not possible yet.

As I understand, you're already detecting the binary path for the authorization prompt? So that much is already possible, though with some known limitations (can't distinguish scripts etc).

Remembering the authorization past client lifetime only requires to store that already-detected binary path somewhere (whether in the DB or only in memory is a separate question). Then on next authorization request, you compare the newly detected path to the list of saved authorized ones.

The only thing that fingerprinting adds to this, is that on top of the binary path, you also save (and compare) some sort of fingerprint of that binary, which gives some additional confidence. The fingerprint can be as simple as just a sha hash of the binary, or you can also include the parent processes tree if you want to be more strict. Just the path alone is already a very very simplistic type of "fingerprint".

So I don't see any technical show stoppers for this, with or without the sha fingerprinting. The known limitations remain as they are, but they're not show-stoppers.

> Something like the following?

How about adding a virtual "Future items" item? That way it will be clearer to both us and the user if that is included or not.

---

One other thing, if you want to show the process tree, I think it should be collapsed by default, since it's rather technical and adds a bunch of visual noise. But allow the user to expand it if they want to examine it.

**Comment by droidmonkey** [comment]
"Remember" implies you are storing the choice somewhere forever. That is what I am talking about with only keep that choice for as long as keepassxc is running. Don't store it in the settings or database. 

Keeping it only in memory reduces the risk and, imo, eliminates the need for fancy fingerprinting.

**Comment by sunjam** [comment]
> Starting Element, has the same behavior, but it only asks for one single entry, not for 10-20 like Remmina. This problem is not specific to Remmina, just with Remmina it is more obvious, because it reads multiple entries in KeePassXC.

Element is working for me.  It is [Nextcloud](https://docs.nextcloud.com/server/latest/developer_manual/client_apis/LoginFlow/index.html) that asks for new authentication every reboot.  _Happy to file a new bug report if that helps; just leaving the mention here for now._

**Comment by droidmonkey** [comment]
There is no need for another issue report.

**Comment by Aetf** [comment]
> So I don't see any technical show stoppers for this, with or without the sha fingerprinting. The known limitations remain as they are, but they're not show-stoppers.

Regardless of binary path or sha fingerprinting, they can't completely solve the problem of python/scripting language.
Whether that is a showstopper or not, let's keep the discussion in #6458.

What I'm saying here is that before we decide on something, all we have right now is remembering as long as the client is
running, and we should base the dialog design on that.

> How about adding a virtual "Future items" item? That way it will be clearer to both us and the user if that is included or not.

That's actually a good idea. Rather than a "Future items" item, maybe an extra "Include future items" check below the list is better?

> One other thing, if you want to show the process tree, I think it should be collapsed by default, since it's rather technical and adds a bunch of visual noise. But allow the user to expand it if they want to examine it.

That's already the case. I just wanted to show what the fully expanded dialog looks like.

> "Remember" implies you are storing the choice somewhere forever. That is what I am talking about with only keep that choice for as long as keepassxc is running. Don't store it in the settings or database.

That is already what we are currently doing (nothing is stored in the settings or database). We only have a text message explaining the decision is remembered while the client is running. It's also remembered while keepassxc is running, but that is sort of implied.

> Keeping it only in memory reduces the risk and, imo, eliminates the need for fancy fingerprinting.

Fingerprinting (or rather any technique to identify a client) is still needed if we want to remember across **client** restart. I.e. the decision is remembered even if Remmina is completely stopped and restarted, while keepassxc is running.

Anyway, whether to remember across keepassxc restarts or across client restarts, those are different issues than what is reported here. What we have now is, that the decision is remembered while the client is running *and* keepassxc is running.

**Comment by michaelk83** [comment]
> Rather than a "Future items" item, maybe an extra "Include future items" check below the list is better?

That does fit the phrasing better. If the prompt is "Client xxx is requesting access to the following entries:", it's not actually asking for the future entries yet.
